src/KeyringController.ts

```diff
@@ -1,41 +1,60 @@
-const encryptor = require('@metamask/browser-passworder');
-const HdKeyring = require('@metamask/eth-hd-keyring');
-const { normalize: normalizeAddress } = require('@metamask/eth-sig-util');
-const SimpleKeyring = require('@metamask/eth-simple-keyring');
+import type { TypedTransaction, TxData } from '@ethereumjs/tx';
+import encryptor from '@metamask/browser-passworder';
+import HDKeyring from '@metamask/eth-hd-keyring';
+import { normalize as normalizeAddress } from '@metamask/eth-sig-util';
+import SimpleKeyring from '@metamask/eth-simple-keyring';
+import type {
+  Hex,
+  Json,
+  Bytes,
+  Keyring,
+  KeyringClass,
+  Eip1024EncryptedData,
+} from '@metamask/utils';
 // TODO: Stop using `events`, and remove the notice about this from the README
 // eslint-disable-next-line import/no-nodejs-modules
-const { EventEmitter } = require('events');
-const ObservableStore = require('obs-store');
+import { EventEmitter } from 'events';
+import ObservableStore from 'obs-store';
+
+import { HEX_PREFIX, KeyringType } from './constants';
+import { State, MessageParams } from './types';
 
 const defaultKeyringBuilders = [
   keyringBuilderFactory(SimpleKeyring),
-  keyringBuilderFactory(HdKeyring),
+  keyringBuilderFactory(HDKeyring),
 ];
 
-const KEYRINGS_TYPE_MAP = {
-  HD_KEYRING: 'HD Key Tree',
-  SIMPLE_KEYRING: 'Simple Key Pair',
-};
-
 /**
  * Strip the hex prefix from an address, if present.
  *
- * @param {string} address - The address that might be hex prefixed.
- * @returns {string} The address without a hex prefix.
+ * @param address - The address that might be hex prefixed.
+ * @returns The address without a hex prefix.
  */
-function stripHexPrefix(address) {
-  if (address.startsWith('0x')) {
+function stripHexPrefix(address: Hex | string): string {
+  if (address.startsWith(HEX_PREFIX)) {
     return address.slice(2);
   }
   return address;
 }
 
 class KeyringController extends EventEmitter {
-  //
-  // PUBLIC METHODS
-  //
+  keyringBuilders: { (): Keyring<Json>; type: string }[];
+
+  public store: typeof ObservableStore;
+
+  public memStore: typeof ObservableStore;
+
+  public encryptor: any;
```

Comment:
Any way we can provide a real type here?

Reply:
Not sure what's the best approach here since `@metamask/browser-passworder` is providing a series of methods. I can try to define a type like this one,

```ts
export type Encryptor = {
  encrypt<R>(
    password: string,
    dataObj: R,
    key?: CryptoKey,
    salt?: string,
  ): Promise<string>;

  decrypt(password: string, text: string, key?: CryptoKey): Promise<unknown>;

  encryptWithDetail<R>(
    password: string,
    dataObj: R,
    salt?: string,
  ): Promise<DetailedEncryptionResult>;

  decryptWithDetail(
    password: string,
    text: string,
  ): Promise<DetailedDecryptResult>;

  encryptWithKey<R>(key: CryptoKey, dataObj: R): Promise<EncryptionResult>;

  decryptWithKey<R>(key: CryptoKey, payload: EncryptionResult): Promise<R>;

  importKey(keyString: string): Promise<CryptoKey>;
};
```

wdyt?

Reply:
`encrypt` and `decrypt` should be the only mandatory functions; those are the ones relied upon by mobile.

The other methods are used only when the `cacheEncyptionKey` option is enabled. If that is enabled they are required, but otherwise we should accept the simpler encryptor to preserve type compatibility with mobile.

Reply:
Hmm, maybe we could make the class generic? So we can ensure the extra methods are required for extension but not for mobile.

Reply:
I have created a ticket to reflect the work to design and develop the data type for `encryptor` https://github.com/MetaMask/KeyringController/issues/206